<a href="https://colab.research.google.com/github/mariawandhare/data_science_66/blob/main/Practical1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Get input from the user
# The input() function gets text from the user.
# int() converts that text into a whole number.
num = int(input("Enter a number: "))

# Convert the number to a string
# This is easier for checking if it's a palindrome
num_str = str(num)

# Reverse the string
# [::-1] is a simple way to create a reversed copy of the string
reversed_num_str = num_str[::-1]

# Check if the original string is the same as the reversed string
if num_str == reversed_num_str:
    print(f"The number {num} is a palindrome.")
else:
    print(f"The number {num} is not a palindrome.")

Enter a number: 121
The number 121 is a palindrome.


### Explanation of the code:

1.  **`num = int(input("Enter a number: "))`**:
    *   This line asks you to type a number. Whatever you type is read as text by `input()`.
    *   `int()` then changes that text into a whole number, which is stored in the variable `num`.

2.  **`num_str = str(num)`**:
    *   To check for a palindrome (which reads the same forwards and backward), it's easiest to work with the number as text (a string).
    *   This line converts the number stored in `num` into a string and stores it in `num_str`.

3.  **`reversed_num_str = num_str[::-1]`**:
    *   This line creates a reversed version of our number string.
    *   `[::-1]` is a special way in Python to create a reversed copy of a sequence (like a string or a list).
    *   So, if `num_str` is "121", `reversed_num_str` becomes "121". If `num_str` is "123", it becomes "321".

4.  **`if num_str == reversed_num_str:`**:
    *   This `if` statement checks if the original number string is exactly the same as its reversed version.
    *   If they are the same, the number is a palindrome.

5.  **`print(f"The number {num} is a palindrome.")`**:
    *   If the condition in the `if` statement is true, this line prints a message saying the number is a palindrome.

6.  **`else:`**:
    *   If the original string is *not* the same as the reversed string, the code inside the `else` block runs.

7.  **`print(f"The number {num} is not a palindrome.")`**:
    *   This line prints a message indicating that the number is not a palindrome.

### Sample Input and Output:

If you enter `121`:

```
Enter a number: 121
The number 121 is a palindrome.
```

If you enter `123`:

```
Enter a number: 123
The number 123 is not a palindrome.
```

In [ ]:
# Get input from the user
# The input() function gets text from the user.
# int() converts that text into a whole number.
num = int(input("Enter a whole number: "))

# Use the modulus operator (%) to check if the number is even or odd.
# If a number divided by 2 has a remainder of 0, it's even.
# Otherwise, it's odd.
if num % 2 == 0:
    # If the remainder is 0, print that the number is even.
    print(f"The number {num} is even.")
else:
    # If the remainder is not 0, print that the number is odd.
    print(f"The number {num} is odd.")

Enter a whole number: 66
The number 66 is even.


### Explanation of the code:

1.  **`num = int(input("Enter a whole number: "))`**:
    *   This line first asks you to type a whole number using the message "Enter a whole number: ".
    *   Whatever you type is read as text by `input()`.
    *   Then, `int()` changes that text into a whole number (an integer), and this number is stored in a variable called `num`.

2.  **`if num % 2 == 0:`**:
    *   This is an `if` statement, which checks if a condition is true.
    *   `num % 2` uses the **modulus operator** (`%`). This operator gives you the remainder after division.
    *   If you divide a number by `2` and the remainder is `0` (meaning `num % 2 == 0` is true), it means the number is even.

3.  **`print(f"The number {num} is even.")`**:
    *   If the condition in the `if` statement is true (the number is even), this line prints a message saying that the number is even.

4.  **`else:`**:
    *   If the condition in the `if` statement is false (the number is not even, meaning it's odd), the code inside the `else` block runs.

5.  **`print(f"The number {num} is odd.")`**:
    *   This line prints a message saying that the number is odd.

### Sample Input and Output:

If you enter `4`:

```
Enter a whole number: 4
The number 4 is even.
```

If you enter `7`:

```
Enter a whole number: 7
The number 7 is odd.
```

In [ ]:
# Get input from the user for the number
# The input() function gets text from the user.
# int() converts that text into a whole number.
num = int(input("Enter a number to display its multiplication table: "))

# Print a header for the multiplication table
print(f"Multiplication Table for {num}:")

# Use a for loop to iterate from 1 to 10
# The range(1, 11) function generates numbers from 1 up to (but not including) 11, so 1 to 10.
for i in range(1, 11):
    # Calculate the product of the input number and the current loop number
    product = num * i
    # Print the multiplication equation in a clear format
    # For example, if num is 5 and i is 2, it will print "5 x 2 = 10"
    print(f"{num} x {i} = {product}")

Enter a number to display its multiplication table: 11
Multiplication Table for 11:
11 x 1 = 11
11 x 2 = 22
11 x 3 = 33
11 x 4 = 44
11 x 5 = 55
11 x 6 = 66
11 x 7 = 77
11 x 8 = 88
11 x 9 = 99
11 x 10 = 110


### Explanation of the code:

1.  **`num = int(input("Enter a number to display its multiplication table: "))`**:
    *   This line first asks you to type a number using the message "Enter a number to display its multiplication table: ".
    *   Whatever you type is first read as text by `input()`.
    *   Then, `int()` changes that text into a whole number (an integer), and this number is stored in a variable called `num`.

2.  **`print(f"Multiplication Table for {num}:")`**:
    *   This line simply prints a title for the multiplication table, making the output clear. For example, if `num` is `7`, it will print "Multiplication Table for 7:".

3.  **`for i in range(1, 11):`**:
    *   This is a `for` loop, which means it will repeat a set of instructions multiple times.
    *   `range(1, 11)` tells the loop to count from `1` up to `10` (it stops before `11`).
    *   In each repetition (or "iteration"), the current number from the count (1, then 2, then 3, and so on, up to 10) is stored in a temporary variable called `i`.

4.  **`product = num * i`**:
    *   Inside the loop, this line calculates the multiplication.
    *   It takes the number you entered (`num`) and multiplies it by the current value of `i` (which goes from 1 to 10).
    *   The result of this multiplication is stored in a variable called `product`.

5.  **`print(f"{num} x {i} = {product}")`**:
    *   This line prints each line of the multiplication table.
    *   The `f"{num} x {i} = {product}"` part is an f-string, which is a simple way to combine text with the values of variables.
    *   For example, if `num` is `7` and `i` is `3`, `product` would be `21`. This line would then print: "7 x 3 = 21".

### Sample Input and Output:

If you enter `7`:

```
Enter a number to display its multiplication table: 7
Multiplication Table for 7:
7 x 1 = 7
7 x 2 = 14
7 x 3 = 21
7 x 4 = 28
7 x 5 = 35
7 x 6 = 42
7 x 7 = 49
7 x 8 = 56
7 x 9 = 63
7 x 10 = 70
```

In [ ]:
# Get input from the user
# The input() function gets text from the user.
# int() converts that text into a whole number.
num = int(input("Enter a non-negative number: "))

# Initialize a variable to store the factorial
# We start with 1 because multiplying by 1 does not change the result.
fact = 1

# Check if the number is negative
# Factorial is not defined for negative numbers.
if num < 0:
    print("Factorial does not exist for negative numbers.")
# Check if the number is 0
# The factorial of 0 is 1.
elif num == 0:
    print("The factorial of 0 is 1")
# Calculate factorial for positive numbers
else:
    # Loop from 1 up to the number (including the number itself)
    for i in range(1, num + 1):
        # Multiply 'fact' by the current number 'i' in each step
        fact = fact * i
    # Print the final result
    print(f"The factorial of {num} is {fact}")

Enter a non-negative number: 6
The factorial of 6 is 720


### Explanation of the code:

1.  **`num = int(input("Enter a non-negative number: "))`**: This line asks you to type a number. Whatever you type is first read as text, and then `int()` changes it into a whole number, which is stored in a variable called `num`.

2.  **`fact = 1`**: We create another variable called `fact` and give it a starting value of `1`. This variable will hold our final factorial result.

3.  **`if num < 0:`**: This checks if the number you entered is less than 0 (a negative number). If it is, the program prints a message because factorials are not defined for negative numbers.

4.  **`elif num == 0:`**: If the number is not negative, this checks if it is exactly `0`. The factorial of `0` is always `1`.

5.  **`else:`**: If the number is positive (not negative and not zero), this part of the code runs.

    *   **`for i in range(1, num + 1):`**: This is a simple loop. It tells the program to count from `1` up to the number you entered (`num`). For each number in this count, it assigns it to a temporary variable `i`.
    *   **`fact = fact * i`**: Inside the loop, in each step, the current value of `fact` is multiplied by `i` (the current number from the count), and the result is stored back into `fact`. So, `fact` keeps track of the running product.
    *   **`print(f"The factorial of {num} is {fact}")`**: After the loop finishes counting through all numbers up to `num`, the program prints the final `fact` value, which is the factorial.

### Sample Input and Output:

If you enter `5`:

```
Enter a non-negative number: 5
The factorial of 5 is 120
```

Because: `1 * 2 * 3 * 4 * 5 = 120`